plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestJdbcConnectionCreation.java

```diff
@@ -0,0 +1,163 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Binder;
+import com.google.inject.Module;
+import com.google.inject.Provides;
+import com.google.inject.Singleton;
+import io.trino.plugin.jdbc.credential.CredentialProvider;
+import io.trino.plugin.jdbc.mapping.IdentifierMapping;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.testing.AbstractTestQueryFramework;
+import io.trino.testing.QueryRunner;
+import org.h2.Driver;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.trino.plugin.jdbc.H2QueryRunner.createH2QueryRunner;
+import static io.trino.plugin.jdbc.TestingH2JdbcModule.createProperties;
+import static io.trino.tpch.TpchTable.NATION;
+import static io.trino.tpch.TpchTable.REGION;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+@Test(singleThreaded = true)
+public class TestJdbcConnectionCreation
+        extends AbstractTestQueryFramework
+{
+    private final Map<Connection, Exception> connectionCreations = new ConcurrentHashMap<>();
+    private final AtomicInteger openConnectionCount = new AtomicInteger();
+    private final AtomicInteger closedConnectionCount = new AtomicInteger();
+
+    @BeforeMethod
+    public void verifyAllConnectionsAreClosed()
+    {
+        Optional<Exception> leakedConnection = connectionCreations.values()
+                .stream()
+                .findFirst();
+        if (leakedConnection.isPresent()) {
+            fail("Connection leaked", leakedConnection.get());
+        }
+        assertThat(openConnectionCount.get()).isEqualTo(closedConnectionCount.get());
+    }
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        return createH2QueryRunner(ImmutableList.of(NATION, REGION), createProperties(), new ReusingConnectionH2Module());
+    }
+
+    @Test
+    public void testOpenConnectionsForShowQueries()
+    {
+        assertOpenConnections("SHOW SCHEMAS", 1);
+        assertOpenConnections("SHOW TABLES", 2);
+    }
+
+    @Test
+    public void testOpenConnectionsForTableQueries()
+    {
+        assertOpenConnections("SELECT * FROM nation", 3);
+        assertOpenConnections("CREATE TABLE copy_of_nation AS SELECT * FROM nation", 10);
+        assertOpenConnections("INSERT INTO copy_of_nation SELECT * FROM nation", 11);
+        assertOpenConnections("DROP TABLE copy_of_nation", 3);
```

Comment:
`...ForTableQueries`  should include SELECT, DELETE, UPDATE and MERGE.

UPDATE and MERGE are not supported, but it's OK to test-cover them here (assert fails ..., don't need to assert the connection counts), so that the test gets updated when these become supported in the future